### ISSUE TITLE
[Bug] keyboard.send_key("<up>") works on desktop but not in the app

### ISSUE THREAD

**Issue by class101** [autokey/autokey#537]
## Version

AutoKey version: aur/autokey-gtk 0.95.10-4

Used GUI (Gtk, Qt, or both): Gtk

Installed via: (PPA, pip3, …). Arch Linux AUR

Linux Distribution: Arch Linux

## Summary

I'm trying to bind Z to keyboard.send_key('<up>'); it works on desktop, but it doesn't work in the Steam game Company Of Heroes 2 

## Steps to Reproduce (if applicable)

- I'm creating a script called `key_z_moveup`
- Pasting this content : `keyboard.send_key('<up>');`
- Hotkey : z
- Window Filter : Company Of Heroes 2

## Expected Results

- Moving up when I press Z when in game Company Of Heroes 2

## Actual Results

- Instead, when I press Z  when in game Company Of Heroes 2, nothing happen, but on desktop (after removing window filter) it works perfectly

Here is the log
```
2021-04-27 19:33:34,365 INFO - service - Matched Script "key_z_moveup" with hotkey and prompt=False
2021-04-27 19:33:34,365 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()

```

But I don't think the log is relevant, since I have another binding `key_tab_nextunit` bound to `keyboard.send_key(";");`, this one works perfectly in CoH2 and with same noise in the logs

```
2021-04-27 19:33:31,565 INFO - service - Matched Script "key_tab_nextunit" with hotkey and prompt=False
2021-04-27 19:33:31,565 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
```

## Notes

I have tried 
*  `keyboard.press_key`
* `keyboard.send_keys`
* `keyboard.send_key`

But unfortunately nothing working with <up> key

**Comment by josephj11** [comment]
TL;DR: switch to `keyboard.send_keys()` and, if that doesn't fix it, submit a trace so we can see what's going on.

Several things to address:

1) Don't use `keyboard.sendkey()` unless you have a specific need for it such as needing its repeat parameter. Always start with `keyboard.send_keys()`. (This might not help with this specific issue, but it will avoid a lot of other issues.)

2) Whenever AutoKey appears to be misbehaving or staring at a script doesn't reveal why it doesn't work, then running a [trace](https://github.com/autokey/autokey-python2/wiki/Problem-Reporting-Guide) often helps and will usually be needed for the developers to figure out bugs anyway.

3) AutoKey generates "synthetic keypresses". I don't know exactly what that means, but some applications, notably terminals, ignore synthetic keypresses by default. Running a trace will show whether AutoKey is doing what it's supposed to, but the application is ignoring it.

Usually, pasting with the clipboard gets around such problems, but that won't work for things like arrow keys.

4) Using the press/release API calls requires doing things differently than what you would expect and is very unlikely to be of use for what you are trying to do here.

5) The locking error message is just noise in this case. It has been there for many years with no consequences and will probably be suppressed in 0.96.

**Comment by class101** [comment]
Hello @josephj11 

Thank you for the quick reply and very detailed response :)

I just did a 

```
strace -p 3649 -e 'trace=!newfstatat,futex,sched_yield,rt_sigprocmask' > ~/strace.log 2>&1
```

[strace.tar.gz](https://github.com/autokey/autokey/files/6388309/strace.tar.gz)

Let me know if you need more informations

About my environment

- The game is running under Proton (wine 6.3)
- Swtched to `keyboard.send_keys`
- And below is log with autokey running from command line
- Also tested with the most up to date latest git commit but no changes

```
2021-04-28 02:23:37,601 DEBUG - service - Input queue at end of handle_keypress: deque(['z'], maxlen=150)
2021-04-28 02:23:37,601 DEBUG - iomediator - Send via event interface
2021-04-28 02:23:37,601 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2021-04-28 02:23:37,602 DEBUG - interface - Send special key: ['<up>']
2021-04-28 02:23:38,016 DEBUG - service - Raw key: 'z', modifiers: [], Key: z
2021-04-28 02:23:38,016 DEBUG - service - Window visible title: 'Company Of Heroes 2', Window class: 'steam_app_231430.steam_app_231430'
2021-04-28 02:23:38,016 INFO - service - Matched Script "key_z_moveup" with hotkey and prompt=False
2021-04-28 02:23:38,017 DEBUG - service - Script runner executing: Script('key_z_moveup')
2021-04-28 02:23:38,017 DEBUG - service - Input queue at end of handle_keypress: deque(['z'], maxlen=150)
2021-04-28 02:23:38,017 DEBUG - iomediator - Send via event interface
2021-04-28 02:23:38,017 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2021-04-28 02:23:38,017 DEBUG - interface - Send special key: ['<up>']
2021-04-28 02:23:38,320 DEBUG - service - Raw key: 'z', modifiers: [], Key: z
2021-04-28 02:23:38,320 DEBUG - service - Window visible title: 'Company Of Heroes 2', Window class: 'steam_app_231430.steam_app_231430'
2021-04-28 02:23:38,320 INFO - service - Matched Script "key_z_moveup" with hotkey and prompt=False
2021-04-28 02:23:38,320 DEBUG - service - Script runner executing: Script('key_z_moveup')
2021-04-28 02:23:38,320 DEBUG - service - Input queue at end of handle_keypress: deque(['z'], maxlen=150)
2021-04-28 02:23:38,320 DEBUG - iomediator - Send via event interface
2021-04-28 02:23:38,320 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2021-04-28 02:23:38,321 DEBUG - interface - Send special key: ['<up>']
```

**Comment by class101** [comment]
Also I notice with autokey-qt, it is slightly different behavior, when I long-press Z, I see the screen is moving up from time to time. It behaves like the game had a lock on it the UP key, then autokey steal the lock, the game recover it back, etc etc

Here is a strace this time with autokey-qt , latest git

[strace.tar.gz](https://github.com/autokey/autokey/files/6388339/strace.tar.gz)

**Comment by class101** [comment]
on sidenote, I tested with 

```
xmodmap -e "keycode 25 = Up"
```

this works great, z correctly trigger UP in the game :)

I don't know if it's technically possible to add an activable support in the options for xmodmap, or a special api that would call xmodmap

**Comment by josephj11** [comment]
Looks like AutoKey did what it was supposed to do, but the app ignored it. AutoKey really doesn't like remapped keyboards, so using xmodmap may end up causing you other problems.

What you can try is to use the xdo Python module or call xdotool directly from your script using the subprocess module. If the problem is the app ignoring synthetic keypresses, this may work.

When I say a "trace", I mean an AutoKey trace as described in the link lower in the previous post. I think an strace is overkill for most things we encounter. Your response above gave us what we needed to see that AutoKey did emit the arrow key.

**Comment by class101** [comment]
Ha sorry joseph for the trace, wasn't aware of this mechanic into the app.

Actually I have found a workaround is to use AutoHotKey through Proton, it works perfectly and I have been able to setup all the following keybinds, especially the one asking mouse click + mouse movements, who gave me problems to setup natively

```
;#InstallKeybdHook
; -----------------------------------
; ------- Company Of Heroes 2 -------
; -----------------------------------
#IfWinActive, Company Of Heroes 2
  Tab::SC033                    ; ';' (next unit selection)
  z::Up                         ; '↑' (camera up)
  q::Left                       ; '←' (camera left)
  s::Down                       ; '↓' (camera down)
  d::Right                      ; '→' (camera right)
  MButton::LAlt                 ; 'LAlt' (camera rotation)
  r::e                          ; A_TESTER
  space::Numpad0                ; A_TESTER

  LAlt::MButton                 ; 'MButton' (camera move)
  ~LButton & ~RButton::MButton  ; 'MButton' (camera move)
  ~RButton & ~LButton::MButton  ; 'MButton' (camera move)
  e::z                          ; A_TESTER
  g::q                          ; A_TESTER
  b::s                          ; A_TESTER
  n::d                          ; A_TESTER
#IfWinActive
```
Will close the ticket but let me know if you need me to test more things, thank you for your time :)

**Comment by josephj11** [comment]
Thanks.
If you have managed to get AutoHotKey to work on Linux, it would be great if you could post on one of our support lists with more details of how to do that. We get a lot of AHK refugees over here and we could point them to it if they are interested in that approach.

Our [support list](https://groups.google.com/forum/#!forum/autokey-users) for end-user support and our [Gitter page](https://gitter.im/autokey/autokey) for more advanced users. 

Our next release, 0.96, will have quite a few enhancements including expanded API calls for mouse events. There is no planned release date for that yet, but it's in the works.

**Comment by class101** [comment]
Sure I can do that, do you think your users are using Proton and are confortable using it ? Because  I have already documented the process I have used to install a Windows application as a non-Steam game for running with Proton (I did it with Vortex Mod Manager as a sample), and I extrapolated the method to install AutoHotkey.

To resume the guide, I show how to setup a Proton prefix with all the necessary Windows redistributables with .NET included. Here I technically use it to setup a single prefix that I link afterwards to every game to manage a single place with all the redist, much like a normal Windows system.

https://steamcommunity.com/sharedfiles/filedetails/?id=2386287653

Maybe I can duplicate it for use with AutoHotkey, technically, the install of AHK didn't cause any trouble with wine, the only trick is to run it with the path to your script as launch paramater to avoid having the .chm opening with hh.exe, autohotkey.chm is not displayed properly under Wine but works under xchm.

**Comment by josephj11** [comment]
Great! Thank you. I have never encountered Proton, so it's unlikely that a lot of our users know about it. But our users don't need it, so that's kind of a moot point.

If there are good tutorials elsewhere, you can just link to them, etc. If it's complicated to setup or a lot of work to document, then you could just provide some hints/bullet points about it because it may interest a pretty small niche of users and so might not justify the effort.

I'm also a bit concerned that if we give this to users, they won't use AutoKey unless they need the extended power of Python scripting or don't want to deal with wine.  But, ultimately, we want to help people get things done, not just convert them to AutoKey.
